### PR TITLE
Fix caret position for errors with utf source

### PR DESCRIPTION
**Pull request by Marenz**
refs #4614

### CODE REVIEW

**Comment by Marenz** [comment]
This case is still faulty, looking into it:
```
2 | /*  2017 */ constructor () public { "©©©©©©©©©"; }
  |                                     ^^^^^^^^^^^^^^^^^^^^

```

File: liblangutil/SourceReferenceFormatterHuman.cpp [hunk 1]
```diff
@@ -103,9 +105,14 @@ void SourceReferenceFormatterHuman::printSourceLocation(SourceReference const& _
 		m_stream << leftpad << ' ';
 		frameColored() << '|';
 		m_stream << ' ';
+
+		std::wstring_convert<std::codecvt_utf8_utf16<wchar_t>> converter;
```
Comment (chriseth) [inline liblangutil/SourceReferenceFormatterHuman.cpp]: Is that really the right fix? Don't we have to compute the byte size in utf-8 encoding?

I think some code in UTF8.cpp can easily be adapted to compute the number of characters represented by a sequence of utf8 bytes.

Comment (Marenz) [reply liblangutil/SourceReferenceFormatterHuman.cpp]: Yeah, I just noticed it's very incomplete. This only works properly if the UTF parts are before the reference location. If it's inside it fails.

Comment (Marenz) [reply liblangutil/SourceReferenceFormatterHuman.cpp]: The `.size()` of a `std::string` is the byte size, of a `std::wstring` it is the amount of code points (as long as the unicode character is within 16 bytes that is)

Comment (Marenz) [reply liblangutil/SourceReferenceFormatterHuman.cpp]: Actually we also can use std::u32string to get the full range. However this doesn't consider grapheme clusters but I think it's not worth fixing it to that extend.

**Comment by Marenz** [comment]
Pushed a probably simpler solution. But the fuzzer can't handle my 4byte unicode test character and fails

File: libsolutil/UTF8.h [hunk 2]
```diff
@@ -38,4 +38,6 @@ inline bool validateUTF8(std::string const& _input)
 	return validateUTF8(_input, invalidPos);
 }
 
+size_t offsetToUTFCodepoint(std::string const& _input, size_t _offset);
```
Comment (chriseth) [inline libsolutil/UTF8.h]: Are you sure you mean `codepoint` here? What does this function return?

Comment (Marenz) [reply libsolutil/UTF8.h]: Ah, the name is ambigious I guess. It returns the amount of codepoints at that byte offset

Comment (Marenz) [reply libsolutil/UTF8.h]: `numBytesToCodepoints`?

Comment (chriseth) [reply libsolutil/UTF8.h]: Can you make it take a string without offset?

Comment (chriseth) [reply libsolutil/UTF8.h]: `size_t numCodepoints(std::string const& _utf8EncodedInput)`?

Comment (chriseth) [reply libsolutil/UTF8.h]: By the way, please check what happens if we start or end the string inside a multi-byte character. I think we do not have to handle that nicely, but it should also not fail.

Comment (Marenz) [reply libsolutil/UTF8.h]: The function only counts the beginning of codepoints, so if we start in the middle, we get 0 and if we end early we still get one.

**Comment by Marenz** [comment]
The functionality in this PR makes the current situation better, but it is not a complete solution to the problem, we completely ignore grapheme clusters and in some fonts some unicode characters are wider than others which breaks our length calculation again (example: 65335: Ｗ). 
However, I don't think it is worth having a complete solution as that would be way too much work. I'd consider this PR as "good enough"

**Review by chriseth — APPROVED** [review]

**Pull request merged**